### PR TITLE
Immudb deployment in dockerswarm

### PR DESCRIPTION
**Pull request by namanmanish**

### CODE REVIEW

File: Docker-Swarm-deployment/single-node/immudb/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM python:3.8-slim
+
+RUN pip install immudb-py
+
+ADD script.py /home/
```
Comment (abhi4578) [inline Docker-Swarm-deployment/single-node/immudb/Dockerfile:5]: ```COPY``` is better than using ```ADD``` here. [refer](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy) @namanmanish

File: Docker-Swarm-deployment/single-node/immudb/script.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+client = ImmudbClient("immudb:3322")
+client.login("immudb","immudb")
+
+client.changePassword("immudb",ADMIN_PASSWORD,"immudb")
+client.login("immudb",ADMIN_PASSWORD)
+
+client.databaseCreate("iudxauth")
+client.databaseCreate("iudxcat")
+client.databaseCreate("iudxrsorg")
```
Comment (abhi4578) [inline Docker-Swarm-deployment/single-node/immudb/script.py:30]: iudxrsorg database, test user, iudx_default user are not needed. 
iudx_rs user has RW permission on iudxrs database. @namanmanish

File: Docker-Swarm-deployment/single-node/immudb/script.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+client.createUser("test",IUDX_TEST_PASSWORD,immudb.constants.PERMISSION_RW,"iudxrs")
+
+client.databaseUse("iudxcat")
+client.sqlExec("CREATE TABLE auditingtable(id VARCHAR NOT NULL, userRole VARCHAR NOT NULL,userID VARCHAR NOT NULL,iid VARCHAR NOT NULL,api VARCHAR NOT NULL,method VARCHAR NOT NULL,time INTEGER NOT NULL,iudxID VARCHAR NOT NULL,PRIMARY KEY id);")
+
+client.databaseUse("iudxauth")
+client.sqlExec("CREATE TABLE table_auditing(id VARCHAR NOT NULL, body VARCHAR NOT NULL,userid VARCHAR NOT NULL,endpoint VARCHAR NOT NULL,method VARCHAR NOT NULL,time INTEGER NOT NULL,PRIMARY KEY id);")
+
+client.databaseUse("iudxrs")
+client.sqlExec("CREATE TABLE auditing(id VARCHAR NOT NULL, userid VARCHAR NOT NULL,api VARCHAR NOT NULL,resourceid VARCHAR NOT NULL,time INTEGER NOT NULL,PRIMARY KEY id);")
```
Comment (abhi4578) [inline Docker-Swarm-deployment/single-node/immudb/script.py:46]: Can you print finally the list of users, databases, tables. So that it confirms the users, dbs and tables are created.

**Comment by abhi4578** [comment]
In docker swarm the immudb container should not expose any of port to outside world. All containers that needs to connect to immudb container  through "<service name>:<port>" over overlay network. Hence the users-db-tables creation script must also run in a container attached to overlay network. Hence a seperate stack file to bring up that container, also it should run once and exits without failure. Something like this : https://github.com/datakaveri/iudx-deployment/blob/master/Docker-Swarm-deployment/single-node/resource-server/database/account-generator.yml

**Comment by abhi4578** [comment]
Morever, the immudb container can run on some node (the node labelled for immudb to be run) and this users creation script. The only way the script connect or other containers to connect is over overlay network!

**Comment by abhi4578** [comment]
The user, db. tables users creation script, docker image can be pushed to docker registry finally.

**Comment by abhi4578** [comment]
@namanmanish  can you look at some docker stack  file best practices?
Some are as follows:
- resource constraints - memory and cpu limit
- including logging config, something similar to this but in stack file https://github.com/datakaveri/iudx-deployment/blob/master/Docker-Swarm-deployment/single-node/infrastructure/files/docker-daemon.json. putting into /etc/docker/daemon.json - all the containers running on that host will be configured using that.  But if someone wants to just run stack file and not the whole docker swarm deployment, this would be useful
 
refer this for writing the stack file, ( compose v3 specs ) https://docs.docker.com/compose/compose-file/compose-file-v3/

**Comment by abhi4578** [comment]
Difference between compose and stack files https://stackoverflow.com/questions/43099408/whats-the-difference-between-a-stack-file-and-a-compose-file @namanmanish  @hackcoderr . This would make things clear.  The stack files ( for deploying in docker swarm) and compose file ( for deploying in single machine, can bring up multiple containers but in same machine) written using compose v3 specs with most of the features common to both the type of deployment . But some are exclusive to docker swarm like replicas etc  which  ```docker stack deploy``` understands  from stack file and not ```docker-compose ``` tool

**Comment by namanmanish** [comment]
We shouldn't allow the container to consume as much memory as the host cause if the host memory is filling up it might through OOME and start killing processes. Any process can be subjected to killing including important processes which might have a thereat of bringing down the system.  With `docker run` command you can specify flags like `--memory` ,`--cpus` etc. 

For swarm you can specify it in the deploy option under limits. 
``` 
resources:
  limits:
    cpus: '0.50'
    memory: 500M
```
          
Here '0.5' means 50% of a single CPU core and the memory limit enforced is 500M

There is one more concept called `reservations` which specifies the basic requirements which should be met in order to spin up the container.

```
reservations:
  cpus: '0.025'
  memory: 20M
```

Please note that reservation should be kept as low as required for the container to spin up.

**Comment by abhi4578** [comment]
Adding node labels and splitting of stack files

**Comment by abhi4578** [comment]
Node labels can be used to constrain immudb container to run specific machine. refer [here](https://docs.docker.com/engine/swarm/services/#placement-constraints). This is useful when we want to place immudb on a certain machine.

We can split the stack files and have three types of deployment  as shown [here](https://github.com/hackcoderr/iudx-deployment/blob/keycloak/Docker-Swarm-deployment/single-node/postgres/README.md). For more info on multiple stack info , refer https://docs.docker.com/compose/extends/#multiple-compose-files . 

Three different way of deploying , why?
1) Deployment one - to cater quick deploy without much regard to best practices, node constraints
2) deployment two - the best way to deploy and how it should be in production
3) deployment three-  add custom config like ports mapped outside, etc  for  development of stack file etc.

**Comment by abhi4578** [comment]
Observe the ,gitignore, https://github.com/hackcoderr/iudx-deployment/blob/keycloak/Docker-Swarm-deployment/single-node/postgres/.gitignore. The resources stack file and custom stack file are not got tracked coz its dependent on environment or  custom config.

**Comment by abhi4578** [comment]
Please have a look at last commit @namanmanish . Its tested and works locally

File: Docker-Swarm-deployment/single-node/immudb/example-immudb-stack.resources.yml [hunk 4]
```diff
@@ -1,9 +1,12 @@
-version: '3.4'
+version: '3.9'
 services:
   immudb:
+    cap_drop:
```
Comment (abhi4578) [inline Docker-Swarm-deployment/single-node/immudb/example-immudb-stack.resources.yml:4]: we can declare in base stack file itself.

File: Docker-Swarm-deployment/single-node/immudb/immudb-stack.yml [hunk 5]
```diff
@@ -21,9 +21,6 @@ services:
         replicas: 1
         restart_policy:
           condition: any
-        placement:
```
Comment (abhi4578) [inline Docker-Swarm-deployment/single-node/immudb/immudb-stack.yml]: Why was this removed? any reason?
